### ISSUE TITLE
[vllm_omni, rollout] BAGEL LoRA rollout generation ~28% slower than non-LoRA due to DiffusionBaseLinearLayerWithLoRA overhead

### ISSUE THREAD

**Issue by zhtmike** [verl-project/verl-omni#248]
## Summary

BAGEL LoRA training rollout (`timing_s/gen`) is **~28% slower** (328s vs 256s) per step compared to non-LoRA training with otherwise identical training scripts. The slowdown is in the **inference/generation phase**, not in weight sync. Root cause analysis points to the vllm-omni `DiffusionBaseLinearLayerWithLoRA` using plain `torch.matmul` instead of vllm's optimized `punica_wrapper`, compounded by CFG doubling forward passes and `enforce_eager` execution preventing CUDA graph fusion.

## Reproduction

**Scripts:**
- LoRA: `examples/flowgrpo_trainer/bagel/run_bagel_pickscore_lora.sh`
- Non-LoRA: `examples/flowgrpo_trainer/bagel/run_bagel_pickscore.sh`

**W&B runs:**
- LoRA: `mikecheung/flow_grpo_open/30xjd535` (bagel_pickscore_lora)
- Non-LoRA: `mikecheung/flow_grpo_open/qopgx61g` (bagel_pickscore)

Both use 4 GPUs, `ROLLOUT_TP=1`, 768 images/step, 15 inference steps, CFG enabled.

### 1. W&B Timing Metrics (mean over 300 steps)

| Metric | LoRA | Non-LoRA | Delta |
|--------|------|----------|-------|
| `timing_s/gen` (rollout generation) | **328.3s** | **256.4s** | **+71.9s (+28%)** |
| `timing_per_image_ms/gen` | 427.4ms | 333.9ms | +93.5ms (+28%) |
| `timing_s/agent_loop/generate_sequences/max` | 321.8s | 249.7s | +72.1s (+29%) |
| `timing_s/update_weights` | 4.2s | 10.4s | -6.2s (**faster**) |
| `timing_s/step` (total) | 549.1s | 584.8s | -35.7s (-6%) |

**Key observation:** Weight sync (`update_weights`) is actually faster in LoRA mode (4s vs 10s). The bottleneck is purely in the rollout **generation** phase.

### 2. Script Configuration Differences

Only these differences are relevant:

```diff
# LoRA script adds LoRA config:
+actor_rollout_ref.model.lora_rank=64
+actor_rollout_ref.model.lora_alpha=128
+actor_rollout_ref.model.lora_dtype=float32
+actor_rollout_ref.model.target_modules="['q_proj_moe_gen',...]"

# Non-LoRA adds FSDP2 strategy (doesn't affect vLLM-omni rollout engine):
+actor_rollout_ref.actor.strategy=fsdp2

# SDE window differs (affects training/trajectory recording, NOT rollout):
  LoRA: sde_window_size=2
  Non-LoRA: sde_window_size=3
```

All rollout config (`num_inference_steps=15`, `n=16`, `noise_level=1.3`, etc.) is **identical**.

### 3. Root Cause: DiffusionBaseLinearLayerWithLoRA Uses Naive Matmul

`DiffusionLoRAManager._replace_layers_with_lora()` (`vllm_omni/diffusion/lora/manager.py:340-436`) replaces target linear layers with `DiffusionBaseLinearLayerWithLoRA` wrappers. Every forward pass through these wrappers incurs extra computation.

**`vllm_omni/diffusion/lora/layers/base_linear.py:69-134`:**
```python
def apply(self, x, bias=None):
    output = self.base_layer.quant_method.apply(self.base_layer, x, bias)  # base matmul
    # ... LoRA matmul:
    delta = (x_flat @ A.t()) @ B.t()
    y_flat[:, offset : offset + slice_size] += delta
    return y_flat.view(original_shape)
```

The class docstring (line 14) explicitly states the performance gap:
> "Override that uses direct torch matmul instead of punica_wrapper."

vLLM's standard AR (language model) LoRA path uses `punica_wrapper` — an optimized CUDA kernel for batched LoRA operations with better memory access patterns. The diffusion path has not been ported to use this optimization.

### 4. CFG Doubles Forward Passes Per Step

BAGEL FlowGRPO uses full-range CFG (`verl_omni/pipelines/bagel_flow_grpo/common.py:25-31`):

```python
BAGEL_FLOWGRPO_CFG_DEFAULTS = {
    "cfg_text_scale": 4.0,
    "cfg_interval": (0.0, 1.0),  # full range
    ...
}
```

Each denoising step runs **two model forward passes** (positive + negative prompt), totaling 30 forward passes per image. The LoRA wrapper overhead is applied on every pass.

### 5. Enforce Eager Prevents CUDA Graph Fusion

The deploy config (`bagel_deploy_config.yaml`) uses `enforce_eager: true` (required for diffusion). Without CUDA graphs, per-layer wrapper overhead is not amortized — each layer's `forward()` → `apply()` incurs full Python call overhead + individual kernel launches.

### 6. Ruled-Out Explanations

**SDE window difference:** `sde_window_size=2` (LoRA) vs `=3` (non-LoRA) only controls how many steps record log-probs. The model still executes 15 forward passes in both cases — the scheduler just skips log-prob tensor allocation outside the window. If anything, LoRA should be slightly *faster* (fewer log-prob computations). The 28% slowdown is in the opposite direction.

**`lora_dtype=float32`:** This config stores training-side LoRA params in fp32. On the rollout side, `LoRAModel.from_lora_tensors(dtype=self.dtype)` casts to bf16 during loading (`verl_omni/utils/vllm_omni/utils.py:94-102`). GPU LoRA matmul runs in bf16, same as non-LoRA.

**Weight sync:** The LoRA fast path in `verl_omni/workers/engine_workers.py:882-925` sends only LoRA tensors via ZMQ IPC. Takes 4.2s vs 10.4s for full-weight sync. Not the bottleneck.

**FSDP1 vs FSDP2:** Only affects the training engine, not the vLLM-omni rollout engine.

## Impact

- Per-image rollout generation is 94ms slower (427ms → 334ms)
- Per-step rollout is 72s slower (328s → 256s)
- Mitigated by faster actor updates (LoRA trains fewer params: 159s vs 237s for `update_actor`), but for generation-bound training (larger `n`, more steps), this overhead becomes material

## Possible Directions

1. **Short-term workaround:** Set `actor_rollout_ref.model.lora.merge=true` to merge LoRA into base weights before sync. Avoids wrapper overhead at the cost of full-weight sync each step — a trade-off to benchmark.

2. **Medium-term fix:** Port `punica_wrapper` support to `DiffusionBaseLinearLayerWithLoRA.apply()` in vllm-omni. This would bring diffusion LoRA inference to parity with AR-model LoRA inference performance.

3. **Profiling confirmation:** Run `torch.profiler` or `nsys` on a single rollout step to confirm the exact time breakdown in `apply()` vs base `quant_method.apply()`.

## Related Code Locations

| File | Lines | Role |
|------|-------|------|
| `vllm_omni/diffusion/lora/layers/base_linear.py` | 69–134 | `apply()` — per-layer LoRA matmul using torch.matmul |
| `vllm_omni/diffusion/lora/manager.py` | 340–436 | `_replace_layers_with_lora()` — replaces original layers with wrappers |
| `verl_omni/utils/vllm_omni/utils.py` | 55–131 | `VLLMOmniHijack` — patches `_load_adapter` for in-memory tensors |
| `verl_omni/workers/engine_workers.py` | 824–984 | `update_weights()` — LoRA fast path (not the bottleneck) |
| `verl_omni/workers/rollout/vllm_rollout/utils.py` | 66–137 | `update_weights_from_ipc()` — rollout-side LoRA receiver |
| `verl_omni/pipelines/bagel_flow_grpo/common.py` | 25–31 | CFG defaults driving double forward passes |

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as QwenImage, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

.

### Expected behavior

.

**Comment by cr-gao** [comment]
Hi, I'd like to work on this. Could you assign it to me? I verified it against current sources (`DiffusionBaseLinearLayerWithLoRA.apply()` upstream, `model.lora.merge` split in `engine_workers.py`) and it holds. My current plan:
1. **Microbench** `apply()` in isolation (rank/batch sweep + `torch.profiler` breakdown) to split the +28% between the extra eager matmuls and Python/launch overhead.
2. **Validate end-to-end + quantify the `merge=true` trade-off** on an SD3.5-Medium FlowGRPO proxy (no hardware for the 4-GPU BAGEL run, so I'll rely on your W&B numbers there), profiling: no-LoRA / LoRA / LoRA+merge, reporting `timing_s/gen` and `timing_s/update_weights`. Post a docs note on when to flip `merge=true`.
3. **Upstream fix** — I'll file a linked issue in vllm-project/vllm-omni and drive it there. Two candidates; which would you prefer?
   - (a) Single-adapter merge-on-load: keep the fast LoRA-only sync, merge ΔW=B·A into base weights at adapter-load time, zero per-forward overhead. Open question: unmerge via base copy (memory) vs subtract-old-delta (bf16 drift over many RL steps).
   - (b) Port `punica_wrapper` into the diffusion path: full parity with AR LoRA and helps multi-LoRA serving, but punica's per-token metadata needs adapting to diffusion batching — larger scope, and likely no win over (a) for single-adapter RL.
Either fix ships with a numerical-consistency check (LoRA vs merged outputs) and speed comparison. If someone's already on this, happy to step back.

**Comment by zhtmike** [comment]
> Hi, I'd like to work on this. Could you assign it to me?

Sure! Welcome solution.

**Comment by cr-gao** [comment]
Microbench results: Isolated `DiffusionBaseLinearLayerWithLoRA.apply()` on a 40-layer chain, dim 3584, bf16, RTX 4090, vllm-omni 0.22.0. Rank 64:

| tokens | wrapper (current) | compiled fused delta | merged into base weights |
|--------|-------------------|----------------------|--------------------------|
| 1024   | +12.4%            | +10.4%               | ≈0%                      |
| 4096   | +19.8%            | +13.0%               | ≈0%                      |
| 16384  | +29.9%            | +19.7%               | ≈0%                      |

The 4K–16K range brackets your +28%. Two findings that refine the root cause:

- Overhead grows with tokens and is nearly rank-independent (16/64/128 within 2pp) — so it's **memory-bandwidth-bound, not launch/Python-bound**: the two LoRA GEMMs are only ~12µs/layer, but `y_flat[slice] = y_flat[slice] + delta` materializes the full-width delta and re-touches the (tokens × out_dim) activation three extra times (profiler shows a Memcpy DtoD per layer from the slice assignment).
- Consequently a fused/compiled eager path recovers only ~1/3 of the gap, and even a punica port would leave an estimated ~12% — while merging ΔW into base weights recovers all of it (a wrapper with an inactive adapter measures +1%, so the merged steady state is free).

This makes direction (a) from my plan (i.e. single-adapter merge-on-load in the engine, keeping the LoRA-only weight sync) the clear fix for the RL case. I'll file the upstream vllm-omni issue with these numbers.

**Comment by cr-gao** [comment]
Upstream fix implemented and validated at vllm-project/vllm-omni#5059. Also found a second, larger cost regime on compiled pipelines (wrappers fragment the torch.compile graph, 2.4x) — SD3.5 e2e: no-LoRA 17.1s / wrappers 37.9s / merge-on-load 17.9s per step, ratio_mean 1±1e-6. Once it lands I'll follow up here with the verl-omni config plumbing.

**Comment by cr-gao** [comment]
Status update so this doesn't sit stale. The upstream fix (vllm-project/vllm-omni#5059,
merge-on-load for diffusion LoRA) has been open since 7/12 with review comments addressed and a
rebase on 7/21, but it hasn't drawn maintainer review since the initial round, so the verl-omni-side
config plumbing I planned here is blocked on it. Happy to discuss the next steps!
